File: backend/app/schemas/ai.py
"""Pydantic schemas for AI endpoints."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


# --- Conversation ---

class AiMessageCreateRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=20000)


class AiMemoryCreateRequest(BaseModel):
    memory_type: str = Field(..., min_length=1, max_length=50)
    content: str = Field(..., min_length=1, max_length=10000)
    metadata: dict | None = None


class AiMemoryUpdateRequest(BaseModel):
    content: str | None = None
    metadata: dict | None = None


class AiActionRejectRequest(BaseModel):
    reason: str | None = None


# --- Responses ---

class AiConversationResponse(BaseModel):
    id: str
    project_id: str
    title: str | None
    created_at: datetime
    updated_at: datetime


class AiMessageResponse(BaseModel):
    id: str
    conversation_id: str
    role: str
    content: str
    tool_calls: dict | list | None = None
    tool_call_id: str | None = None
    token_count: int | None = None
    created_at: datetime


class AiConversationDetailResponse(BaseModel):
    id: str
    project_id: str
    title: str | None
    messages: list[AiMessageResponse]
    created_at: datetime
    updated_at: datetime


class AiImpactMetricResponse(BaseModel):
    metric: str
    current: float
    projected: float
    change_pct: float
    direction: str


class AiImpactSummaryResponse(BaseModel):
    affected_metrics: list[AiImpactMetricResponse]
    data_affected_pct: float
    confidence: str


class AiActionResponse(BaseModel):
    id: str
    project_id: str
    conversation_id: str | None
    action_type: str
    title: str
    description: str | None
    reasoning: str | None
    payload: dict
    status: str
    impact_summary: AiImpactSummaryResponse | None = None
    resolved_by: str | None
    resolved_at: datetime | None
    created_at: datetime


class AiMemoryResponse(BaseModel):
    id: str
    project_id: str
    memory_type: str
    content: str
    metadata: dict | None
    created_at: datetime
    updated_at: datetime


class AiHealthIssueResponse(BaseModel):
    severity: str
    category: str
    message: str
    suggested_action: str


class AiHealthResponse(BaseModel):
    id: str
    project_id: str
    health_score: float
    summary: str
    issues: list[AiHealthIssueResponse]
    metrics: dict
    created_at: datetime


class AiConfigureRequest(BaseModel):
    llm_api_key: str | None = Field(None, min_length=1, max_length=500)
    llm_provider: str | None = Field(None, min_length=1, max_length=50)
    llm_model: str | None = Field(None, min_length=1, max_length=100)
    llm_base_url: str | None = None


class AiStatusResponse(BaseModel):
    ai_enabled: bool
    llm_provider: str | None = None
    llm_model: str | None = None
    has_api_key: bool = False
    connected: bool = False

File: frontend/src/types/ai.ts
/* AI type definitions for GoKaatru v2 */

export interface AiConversation {
  id: string;
  project_id: string;
  title: string | null;
  created_at: string;
  updated_at: string;
}

export interface AiMessage {
  id: string;
  conversation_id: string;
  role: "user" | "assistant" | "tool" | "system";
  content: string;
  tool_calls?: unknown;
  tool_call_id?: string | null;
  token_count?: number | null;
  created_at: string;
}

export interface AiConversationDetail extends AiConversation {
  messages: AiMessage[];
}

export interface AiImpactMetric {
  metric: string;
  current: number;
  projected: number;
  change_pct: number;
  direction: string;
}

export interface AiImpactSummary {
  affected_metrics: AiImpactMetric[];
  data_affected_pct: number;
  confidence: string;
}

export interface AiAction {
  id: string;
  project_id: string;
  conversation_id: string | null;
  action_type: string;
  title: string;
  description: string | null;
  reasoning: string | null;
  payload: Record<string, unknown>;
  status: "pending" | "approved" | "rejected" | "failed";
  impact_summary: AiImpactSummary | null;
  resolved_by: string | null;
  resolved_at: string | null;
  created_at: string;
}

export interface AiMemory {
  id: string;
  project_id: string;
  memory_type: string;
  content: string;
  metadata: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface AiHealthIssue {
  severity: "critical" | "warning" | "info";
  category: string;
  message: string;
  suggested_action: string;
}

export interface AiHealth {
  id: string;
  project_id: string;
  health_score: number;
  summary: string;
  issues: AiHealthIssue[];
  metrics: Record<string, unknown>;
  created_at: string;
}

export interface AiStatus {
  ai_enabled: boolean;
  llm_provider: string | null;
  llm_model: string | null;
  has_api_key: boolean;
  connected: boolean;
}
